mlflow/models/evaluation/base.py

```diff
@@ -0,0 +1,530 @@
+        :return: An `EvaluationResult` instance containing evaluation results.
+        """
+        raise NotImplementedError()
+
+
+def list_evaluators():
+    """
+    Return a name list for all available Evaluators.
+    """
+    # import _model_evaluation_registry inside function to avoid circuit importing
+    from mlflow.models.evaluation.evaluator_registry import _model_evaluation_registry
+
+    return list(_model_evaluation_registry._registry.keys())
+
+
+class StartRunOrReuseActiveRun:
+    """
+    A manager context return:
+     - If there's an active run, return the active run id.
+     - otherwise start a mflow run with the specified run_id.
+    """
+
+    def __init__(self, run_id):
+        self.user_specified_run_id = run_id
+        self.managed_run = None
+
+    def __enter__(self):
+        if mlflow.active_run() is not None:
+            active_run_id = mlflow.active_run().info.run_id
+            if (
+                self.user_specified_run_id is not None
+                and self.user_specified_run_id != active_run_id
+            ):
+                raise ValueError(
+                    "An active run exists, you cannot specify another run_id when " "evaluating."
+                )
+            return active_run_id
+        else:
+            if self.user_specified_run_id is None:
+                raise ValueError(
+                    "Active run does not exist, you need specify a run_id when " "evaluating."
+                )
+            self.managed_run = mlflow.start_run(run_id=self.user_specified_run_id).__enter__()
+            return self.user_specified_run_id
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.managed_run is not None:
+            return self.managed_run.__exit__(exc_type, exc_val, exc_tb)
+
+
+def evaluate(
+    model: Union[str, "mlflow.pyfunc.PyFuncModel"],
+    model_type,
+    dataset,
+    run_id=None,
+    evaluators=None,
+    evaluator_config=None,
+) -> "mlflow.models.evaluation.EvaluationResult":
+    """
+    Evaluate a pyfunc model on specified dataset, log evaluation results (metrics and
+    artifacts) into active run or specified mlflow run), and return evaluation results
+    containing metrics and artifacts.
+
+    :param model: A pyfunc model instance, or a URI referring to such a model.
+
+    :param model_type: A string describing the model type. The default evaluator
+                       supports "regressor" and "classifier" as model types.
+    :param dataset: An instance of `EvaluationDataset` containing features
+                    labels (optional) for model evaluation.
+    :param run_id: The ID of the MLflow Run to which to log results. If
+                   unspecified, behavior depends on the specified `evaluator`.
+                   When `run_id` is unspecified, the default evaluator logs
+                   results to the current active run, creating a new active run if
+                   one does not exist.
```

Comment:
What should happen if `run_id` is specified, but there is no active run?

Reply:
Then the specified run should be resumed, assuming it already exists. If it does not exist, we should throw.

Reply:
Thanks for the explanation!